### PR TITLE
Standardize Tuner subclassing

### PR DESCRIPTION
**Pull request by omalleyt12**
Previously, `BaseTuner` users could do:

```python
model = tuner.hypermodel.build(trial.hyperparameters)
```

But `Tuner` users had to do:

```python
model = tuner._build_model(trial.hyperparameters)
tuner._compile_model(model)
```

This PR makes it so that both can just call `tuner.hypermodel.build`, and Tuner users still have the features of being able to override things in compile and to create a model with `tf.distribute`

Also, cleaned up `run_trial` so that it is more clear what is happening with the callbacks and how the results are being reported to the `Oracle` and the models are being saved.

Also, this removes the error if a Model is not compiled, since `model.fit` already handles that, and users of custom training loops don't have to compile their Models.

### CODE REVIEW

File: kerastuner/engine/hypermodel.py [hunk 1]
```diff
@@ -32,3 +42,112 @@ class DefaultHyperModel(HyperModel):
     def __init__(self, build, name=None, tunable=True):
         super(DefaultHyperModel, self).__init__(name=name)
         self.build = build
+
+
+class KerasModelWrapper(HyperModel):
```
Comment (fchollet) [inline kerastuner/engine/hypermodel.py]: This functionality should be factored as overridable methods on the Tuner, not in a separate wrapper class (that can't be overridden), so as to enable users to override and to reduce indirection.

`build` isn't very likely to overridden, but `compile` is quite likely to be overridden (one example would be making the loss a hyperparameter).

File: tests/kerastuner/engine/tuner_workflows_test.py [hunk 2]
```diff
@@ -178,9 +180,9 @@ def test_override_compile(tmp_dir):
         directory=tmp_dir)
 
     assert tuner.oracle.objective.name == 'val_mse'
-    assert tuner.optimizer == 'rmsprop'
-    assert tuner.loss == 'sparse_categorical_crossentropy'
-    assert tuner.metrics == ['mse', 'accuracy']
+    assert tuner.hypermodel.optimizer == 'rmsprop'
```
Comment (fchollet) [inline tests/kerastuner/engine/tuner_workflows_test.py]: You'd expect the class that was passed `optimizer` etc to hold these attributes, e.g. `x = X(foo=bar); x.foo == bar`

File: tests/kerastuner/engine/tuner_workflows_test.py [hunk 3]
```diff
@@ -152,7 +153,8 @@ def test_hypermodel_with_dynamic_space(tmp_dir):
         executions_per_trial=3,
         directory=tmp_dir)
 
-    assert tuner.hypermodel == hypermodel
+    # tuner.hypermodel is a KerasModelWrapper
```
Comment (fchollet) [inline tests/kerastuner/engine/tuner_workflows_test.py]: This is likely to be confusing for anyone accessing `tuner.hypermodel`

File: kerastuner/engine/hypermodel.py (hunk 1 above)
Comment (omalleyt12) [reply kerastuner/engine/hypermodel.py]: IMO conceptually, the functionality provided here (specifying a max model size, allowing failed model builds, using distribution strategies, overriding loss, metrics, etc) is part of building the Model. Having this functionality be part of a HyperModel class buys us the fact that `hypermodel.build` works as expected. Previously, the example below would ignore compiled overrides from the Tuner and would not work with distribution strategies. IMO users should always be using `self.hypermodel.build` to build their Models.

```python
def run_trial(self, trial, x, y):
  model = self.hypermodel.build(trial.hyperparameters)
```

Making loss a HyperParameter is still doable like this:

```python
def my_build_model(hp):
  model = build_model(hp)
  model.compile('adam', loss=hp.Choice('my_loss', ['bce', 'mse']))
  return model
```

A future convenience feature we could add is accepting a `HyperParameter` object for the `Tuner` arguments `loss`, `optimizer`, `metrics`

File: tests/kerastuner/engine/tuner_workflows_test.py (hunk 2 above)
Comment (omalleyt12) [reply tests/kerastuner/engine/tuner_workflows_test.py]: In general agreed but we're already breaking that assumption for our convenience `Tuner`s which accept args that are passed to the `Oracle`. 

We could optionally mirror these here, but if so we should make them properties that refer to the `KerasModelWrapper` properties so that changing them changes the corresponding values in the `KerasModelWrapper`. WDYT?

File: kerastuner/engine/hypermodel.py (hunk 1 above)
Comment (omalleyt12) [reply kerastuner/engine/hypermodel.py]: This is especially an issue because the workflow of recreating the best Model and retraining currently requires the use of private methods:

```python
hps = tuner.get_best_hyperparameters()[0]
model = tuner._build_model(hps)
tuner._compile_model(model)
```

Versus in this PR:

```python
hps = tuner.get_best_hyperparameters()
model = tuner.hypermodel.build(hps)
```

File: tests/kerastuner/engine/tuner_workflows_test.py (hunk 3 above)
Comment (omalleyt12) [reply tests/kerastuner/engine/tuner_workflows_test.py]: We already do some wrapping (turning callables into HyperModel instances), I think while it could be confusing it is less confusing than having `tuner.hypermodel.build` not work properly (especially given the concerns in the first comment above)

**Comment by omalleyt12** [comment]
Thanks for the review!

File: kerastuner/engine/hypermodel.py (hunk 1 above)
Comment (fchollet) [reply kerastuner/engine/hypermodel.py]: That makes sense.

Naming note: this class does not wrap a Keras Model, it wraps a HyperModel. It's also a HyperModel subclass itself. Perhaps `KerasHyperModel` would be a better name (conveys that this is a Keras specialization of the generic HyperModel).

File: kerastuner/engine/tuner.py [hunk 4]
```diff
@@ -89,26 +84,24 @@ def __init__(self,
                  logger=None,
                  tuner_id=None,
                  overwrite=False):
+
+        hypermodel = hm_module.KerasModelWrapper(
```
Comment (fchollet) [inline kerastuner/engine/tuner.py]: Should we expose the class in the public API, and not wrap what was passed by the user if it's already a subclass of the Keras HyperModel? This would allow people to provide their own subclass when necessary.

File: kerastuner/engine/hypermodel.py (hunk 1 above)
Comment (omalleyt12) [reply kerastuner/engine/hypermodel.py]: Done, changed to `KerasHyperModel`

File: kerastuner/engine/tuner.py (hunk 4 above)
Comment (omalleyt12) [reply kerastuner/engine/tuner.py]: Done. Made it so that subclasses of `KerasHyperModel` are not wrapped. I think it's probably ok to leave this as `kt.engine.hypermodels.KerasHyperModel` rather than expose more publicly because the assumptions made by this class are very minimal and so subclassing will probably be niche

The only assumptions made are:

* A Keras model should be returned
* 5 tries at building the model should be allowed before crashing

Even the assumption that the Keras model be compiled is removed in this PR, since `fit` will already check this and users of custom training loops will not expect to need to compile

**Comment by omalleyt12** [comment]
Thanks for the review!

**Comment by omalleyt12** [comment]
Thanks for the review!
